### DISCUSSION TITLE
Which architectures should be included?

### DISCUSSION THREAD

**Discussion by LukeWood** [keras-team/keras-cv#52] [General]
Hey KerasCV contributors,

I'm starting this discussion thread as a place to throw out ideas for which model architectures should/shouldn't be included in KerasCV.  The answer for some architectures feel more obvious.  I.e. RegNets, ResNets, WideResNet, all feel like strong includes.  Some are much less obvious.

Feel free to post links, ideas, or thoughts relating to model architectures in this discussions.

**Comment by bhack** [comment]
Also if this Is probably too fresh It could be interesting to collect some strong self-supervised performer like [ReLICv2](https://arxiv.org/abs/2201.05119)

I also want to try to figure out where we will place common reusable components between Keras-cv and keras-nlp on the emerging thread of multimodal/unified models like:

https://ai.facebook.com/blog/the-first-high-performance-self-supervised-algorithm-that-works-for-speech-vision-and-text/

Or 

https://blog.google/products/search/introducing-mum/

**Comment by innat** [comment]
(IMO) Some arch. that should be included are as follows (**currently the following models are not included in `keras.applications`**).

**Image Classification**
- [ResNeSt](https://github.com/keras-team/keras/issues/15816) - [ResNeXt](https://github.com/keras-team/keras/issues/15270) - [SE-Net, SE-ResNeXt](https://github.com/keras-team/keras/issues/15267) - [SE-ResNet](https://github.com/keras-team/keras/issues/15268) - [NFNets](https://github.com/keras-team/keras/issues/15229) - [ResNet-RS](https://github.com/keras-team/keras/issues/15780) - [GhostNet](https://arxiv.org/abs/1911.11907) - [MixNet](https://arxiv.org/abs/1907.09595) - [ConvNeXt](https://arxiv.org/abs/2201.03545)
- [Swin-Transformer](https://github.com/keras-team/keras/issues/15545) - [DeiT](https://arxiv.org/abs/2012.12877) - [ViT](https://arxiv.org/abs/2010.11929)

**Object Detection** 
- [EfficientDet](https://arxiv.org/abs/1911.09070) - Y0LO ([v3](https://pjreddie.com/darknet/yolo/), [v4](https://arxiv.org/abs/2004.10934), [R](https://github.com/WongKinYiu/yolor), [X](https://github.com/Megvii-BaseDetection/YOLOX)), [Faster-RCNN](https://arxiv.org/abs/1506.01497), [DE⫶TR](https://github.com/facebookresearch/detr), [RetinaNet](https://github.com/fizyr/keras-retinanet), SSD
- [Pix-Seq: A general framework for turning RGB pixels into semantically meaningful sequences](https://github.com/google-research/pix2seq)

![image](https://user-images.githubusercontent.com/17668390/165343313-bd4a1964-0e24-45cd-b944-d5afa9334f8b.png)

**Semantic Segmentation** (backbone: image classification model)

- UNet, [XNet](https://github.com/MrGiovanni/UNetPlusPlus), [DeepLab](https://github.com/tensorflow/models/tree/master/research/deeplab) (v3, v3++), [DANet](https://github.com/junfu1115/DANet), [CCNet](https://github.com/speedinghzl/CCNet), [SegFormer](https://github.com/NVlabs/SegFormer), [SETR](https://arxiv.org/abs/2012.15840), [PointRend](https://github.com/facebookresearch/detectron2/tree/main/projects/PointRend)

**Instance and Panoptic Segmentation**
- [Mask-RCNN](https://arxiv.org/abs/1703.06870)
- [Mask2Former](https://github.com/facebookresearch/Mask2Former)
- more [list](https://www.kaggle.com/c/sartorius-cell-instance-segmentation/discussion/278883#1546104).

**Reply by DavidLandup0** [reply]
Great list! `keras.applications` is lagging behind `torchvision`, so `keras_cv.models` would be a great contender. `torchvision` has a few out-of-the-box object detection classes as well, including RetinaNet and Faster R-CNN that produce results in ~25 lines of code. Their support is pretty flimsy though, and it's still in beta, so there's a clear void for OD model support in official libraries. Perhaps KerasCV would be the first one to bring OD to the masses? :)

**Comment by sebastian-sz** [comment]
Apart from standard architectures, It would be great to see some smaller architectures, that are more suitable for mobile as well:
* `ResNet18` and `ResNet34` - these are [unavailable](https://github.com/keras-team/keras/issues/15269) in Keras ecosystem as far as I know.
* `EfficientNetLite` - not mentioned in the paper, but available in [original repository](https://github.com/tensorflow/tpu/tree/master/models/official/efficientnet/lite) and via TFLite [Model Maker](https://www.tensorflow.org/lite/tutorials/model_maker_image_classification#detailed_process).

**Reply by LukeWood** [reply]
Yeah, we have a contributor working on getting ResNet18 added already.  ResNet34, noted.  I'll open an issue.   This is definitely an add.

How does efficient net lite differ from those offered in Keras applications?

**Reply by innat** [reply]
@LukeWood cc. @sebastian-sz 
Check [this](https://github.com/tensorflow/tpu/tree/master/models/official/efficientnet/lite#efficientnet-lite) page. Mainly stated: 


> Due to the requirements from edge devices, we mainly made the following changes based on the original EfficientNets.
> 
> Remove squeeze-and-excite (SE): SE are not well supported for some mobile accelerators.
> Replace all swish with RELU6: for easier post-quantization.
> Fix the stem and head while scaling models up: for keeping models small and fast.

**Comment by innat** [comment]
A pointer to few **OCR** (text detection-recognition) models.

- [ABCNet](https://github.com/Yuliang-Liu/bezier_curve_text_spotting)
- [Mask TextSpotter v3](https://github.com/MhLiao/MaskTextSpotterV3)
- [CRAFT](https://github.com/clovaai/CRAFT-pytorch)
- [CharNet](https://github.com/msight-tech/research-charnet)
- [keras-io-code_example](https://keras.io/examples/vision/captcha_ocr/)

**Comment by innat** [comment]
@LukeWood Is there any possibility to support 3D modeling? It becomes obvious to have for example in deep-learning-in-medicine practice. FYI, there's some unofficial support, [segmentation_models_3D](https://github.com/ZFTurbo/segmentation_models_3D) and [classification_models_3D](https://github.com/ZFTurbo/classification_models_3D). Here is one nice toolbox for PyTorch, [MONAI](https://monai.io/).

**Reply by LukeWood** [reply]
For now, probably not.  It would be really cool though.  We have limited resources and our first tasks we are targeting are:
- image classification
- object detection
- image segmenetation

**Comment by bhack** [comment]
Some modern `mixer` style networks could be interesting:

https://github.com/locuslab/convmixer

https://discuss.tensorflow.org/t/research-mlp-mixer-an-all-mlp-architecture-for-vision/1849/7

**Reply by innat** [reply]
https://keras.io/examples/vision/convmixer/

**Comment by chjort** [comment]
It would be great to include models from each architectural pillar in deep learning CV. Examples are, but not limited to:
* Convolution based architectures
* Transformer based architectures
* MLP based architectures.
* Hybrid architectures

To begin with, I think it would be good to identify what corresponds to ___the___ ResNet for each pillar. I.e. what is the most known or widely-used model within that type of architecture, exactly like what ResNet is for convolution based architectures. I think it would be great to implement these "core" models to begin with, and then only implement other variants that has a significant level of incremental changes or performance increases.

**Reply by bhack** [reply]
I think that other then a [global citation threshold](https://github.com/keras-team/keras-cv/issues/34#issue-1098232396) we need to add the time dimension. We need to find a metric about citations in the last one or 2 years.
E.g. relu activation paper has 996 total citations, 36 in 2022), 552 in 2021 (Google scholar) 

Resources are limited, however, also if we choice as a policy to scale the repo maintainership over the community and partially rely on its own codeownership/PR reviews capabilities the research in this domain has still a growing output.

I think here it is also a little bit harder as here we are not oriented just to network reproduction but to identify layers, ops/utils, losses, metrics, optimizers and any other kind of component unit that could be reused across many experiments/network flavors.

So, in theory, components that are used only in one o few model with only a marginal improvement on the performance (computational, datasets metrics or both) could potentially have the lowest cost/value ratio for a library to allocate a review and maintainership resources over time.

**Reply by chjort** [reply]
> I think here it is also a little bit harder as here we are not oriented just to network reproduction but to identify layers, ops/utils, losses, metrics, optimizers and any other kind of component unit that could be reused across many experiments/network flavors.

I agree that we need to decide on how sub-components of a model should be reusable and to which degree. Making a component highly generic and reusable requires well thought out abstractions, and can lead to high cognitive load for the user. This is also why the HuggingFace Transformer library specificly does not aim to be modular ([See here](https://github.com/huggingface/transformers#why-shouldnt-i-use-transformers)). They instead optimize towards making each model implementation explicit and easily accessible. Personally, I would like more modularity and abstractions than what Huggingface does, but not too much either. We need to strike a balance.

> So, in theory, components that are used only in one o few model with only a marginal improvement on the performance (computational, datasets metrics or both) could potentially have the lowest cost/value ratio for a library to allocate a review and maintainership resources over time.

This is definitely a metric we need to consider when deciding on the level of abstraction and modularity of a component.

**Comment by innat** [comment]
@LukeWood wdyt? 

[Introducing TorchVision’s New Multi-Weight Support API](https://pytorch.org/blog/introducing-torchvision-new-multi-weight-support-api/) 
[Discussion. ](https://github.com/pytorch/vision/issues/5088)

**Reply by innat** [reply]
[Pretrained EfficientNet Checkpoints](https://github.com/tensorflow/tpu/tree/master/models/official/efficientnet#2-using-pretrained-efficientnet-checkpoints)

**Comment by innat** [comment]
If a project doesn't have peer-reviewed paper but consistently making strong impact to the community, will it be considered to add? For example, model like [yolo-v5](https://github.com/ultralytics/yolov5), [yolo-v6](https://github.com/meituan/YOLOv6), [yolo-v7](https://github.com/jinfagang/yolov7) doesn't have paperwork but very impactful. 

(I wonder why such practice even gets accepted by the community, they should have paperwork first, IMO).

**Reply by LukeWood** [reply]
> If a project doesn't have peer-reviewed paper but consistently making strong impact to the community, will it be considered to add? For example, model like [yolo-v5](https://github.com/ultralytics/yolov5), [yolo-v6](https://github.com/meituan/YOLOv6), [yolo-v7](https://github.com/jinfagang/yolov7) doesn't have paperwork but very impactful.
> 
> (I wonder why such practice even gets accepted by the community, they should have paperwork first, IMO).

Yes absolutely.

**Reply by zhiqwang** [reply]
Maybe YOLOv7 should be this one : https://github.com/WongKinYiu/yolov7, The other one does not seem to be accepted by the community.

Just IMO, YOLOv5 is great, you can find many places in YOLO {R, X} that are affected by YOLOv5.

**Comment by innat** [comment]
There is a nice code repository maintaining regarding [**dense pixel labeling tasks**](https://github.com/google-research/deeplab2) from google research. It would be great if it's shifted towards here in `keras_cv`.

Supported methods / models: 
<img src="https://user-images.githubusercontent.com/17668390/178908745-2747f68f-2ef2-401d-a7a8-3dc25089d92b.png" width="200" height="200" />

cc. @csrhddlam @yucornetto @YknZhu

**Comment by innat** [comment]
[Info]

 Towards Grand Unification of Object Tracking
https://github.com/masterbin-iiau/unicorn

<img width="1146" alt="Unicorn" src="https://user-images.githubusercontent.com/17668390/180932625-892912ac-91f4-44a4-bde6-85f5fdb75cca.png">

**Comment by bhack** [comment]
https://arxiv.org/abs/2208.10442

![image](https://user-images.githubusercontent.com/1710528/186181681-621c058a-6441-4219-b4f8-e91684e38896.png)

@LukeWood I think that we need to talk also with Keras-NLP on where/how to handle these models trend.
